### PR TITLE
fix: Allow owning the same port with different protocols

### LINKED ISSUE #5575: Vector 0.11 cannot bind to both TCP and UDP on same port.
### Vector Version
```
vector 0.10.0 (g0f0311a x86_64-unknown-linux-gnu 2020-07-22)
```

### Vector Configuration File

```
[sources.syslog_tcp_511]
  type = "syslog" # required
  address = "0.0.0.0:511" # required, required when mode = "tcp" or mode = "udp"
  mode = "tcp" # required

[sources.syslog_udp_511]
  type = "syslog" # required
  address = "0.0.0.0:511" # required, required when mode = "tcp" or mode = "udp"
  mode = "udp" # required
```

### Debug Output

```
 vector[1466290]: Dec 17 12:08:55.176 ERROR vector::cli: Configuration error. error=Resource `0.0.0.0:511` is claimed by multiple components: {"syslog_tcp_511", "syslog_udp_511"}
```

### Expected Behavior
Vector running 

### Actual Behavior

Vector can't start because of the above error

### PR DESCRIPTION
**Pull request by juchiast**
close #5575

### CODE REVIEW

**Comment by juchiast** [comment]
Failres look unrelated:
```
2020-12-19T08:37:30.8695840Z failures:
2020-12-19T08:37:30.8696330Z     includes::tests::type_def::value_non_empty_array_infallible
```

File: src/config/mod.rs [hunk 1]
```diff
@@ -298,13 +298,27 @@ inventory::collect!(TransformDescription);
...
 #[derive(Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
 pub enum Resource {
-    Port(SocketAddr),
+    Port(SocketAddr, Protocol),
     SystemFdOffset(usize),
     Stdin,
     DiskBuffer(String),
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
```
Comment (fanatid) [inline src/config/mod.rs:307]: `Copy` can be added here so we will not need `.clone()` call.

**Review by ktff — APPROVED** [review]
LGTM, there are just a few non blocking issues.

File: src/config/mod.rs [hunk 2]
```diff
@@ -420,16 +434,19 @@ impl Resource {
...
+            Protocol::Tcp => write!(fmt, "tcp"),
+        }
     }
 }
 
 impl Display for Resource {
     fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), fmt::Error> {
         match self {
-            Resource::Port(address) => write!(fmt, "{}", address),
+            Resource::Port(address, protocol) => write!(fmt, "{} {}", protocol, address),
```
Comment (jszwedko) [inline src/config/mod.rs:449]:
```suggestion
            Resource::Port(address, protocol) => write!(fmt, "{}:{}", protocol, address),
```

**Review by jszwedko — APPROVED** [review]
Nice work!

**Pull request merged**